### ISSUE TITLE
How to place 16-bit local variable on the stack at -0x1 offset?

### ISSUE THREAD

**Issue by vovaok** [NationalSecurityAgency/ghidra#2383]
I have some code for HCS-08 microcontroller. It has 8-bit big endian CPU. Inside a function the stack frame looks like this:
offset: -0x1    0x0    0x1    0x2    0x3    0x4
          (local_16bit)  (ret_addr)  (param_16bit)
But I can't assign data type "word" to `local_16bit` variable, because "No new data type in the cycle group fits".
I solve the issue temporarily by declaring two 8-bit variables at -0x1 and 0x0 offsets, but decompilation quality is poor.
![image](https://user-images.githubusercontent.com/62897758/96590875-37f1a880-12ef-11eb-8aad-2d6ae7e06507.png)
Also I tried to shift the return address in the .cspec file:
    <returnaddress>
      <varnode space="stack" offset="1" size="2"/>
    </returnaddress>
But this action has no effect.

**Comment by esaulenka** [comment]
Your problem may have following cause:

There are two types of push/pop behavior.
In x86, SP points to last occupied place: PUSH is *(--SP)=reg, POP is reg=*(SP++)
In HCS08, SP points to first free space: PUSH is *(SP--)=reg, POP is reg=*(++SP)

But i didn't find, how this offet indicated in cspec files.

**Comment by vovaok** [comment]
> Your problem may have following cause:
> 
> There are two types of push/pop behavior.
> In x86, SP points to last occupied place: PUSH is _(--SP)=reg, POP is reg=_(SP++)
> In HCS08, SP points to first free space: PUSH is _(SP--)=reg, POP is reg=_(++SP)
> 
> But i didn't find, how this offet indicated in cspec files.

Yeah, you're right. And do you have any ideas how to explain it to GHIDRA? Is it a bug or I can adjust stack frame offset somehow?

**Comment by astrelsky** [comment]
```xml
<stackpointer register="name" space="space" growth="positive/negative" reversejustify="true/false"/>
```

**Comment by vovaok** [comment]
Stackpointer is defined as it should.
The problem is that GHIDRA assumes the return address is at zero offset and locals may have negative offsets only. But at the same time I can define a local variable at any offset. The issue is I can't define multi-byte variable at offset -0x1. It seems that there is the boundary between negative and zero offsets and local variable can't overlay it. So I want to move the boundary between 0x0 and 0x1 offsets. Is there a way to perform this?

**Comment by astrelsky** [comment]
> Stackpointer is defined as it should.
> The problem is that GHIDRA assumes the return address is at zero offset and locals may have negative offsets only. But at the same time I can define a local variable at any offset. The issue is I can't define multi-byte variable at offset -0x1. It seems that there is the boundary between negative and zero offsets and local variable can't overlay it. So I want to move the boundary between 0x0 and 0x1 offsets. Is there a way to perform this?

Oh. My apologies I misunderstood.

**Comment by esaulenka** [comment]
@astrelsky Andrew, I think this settings is not enough. Thanks for doc - I never seen it before (where it can  be found in readable form? I only found [raw xml](https://raw.githubusercontent.com/NationalSecurityAgency/ghidra/49c2010b63b56c8f20845f3970fedd95d003b1e9/Ghidra/Features/Decompiler/src/main/doc/cspec.xml) ).
As I understand sources, `reversejustify` applies only for variables, less than single stack change.

HCS8 (and STM8, which I am currently working on) doesnt have this issue - register size always equal to stack change. 
It seems that to describe correctly its stack feature, one needs to specify non equal stackshift and extrapop values in function prototype. I tried to do this, but failed.

Unfortunately, there is not enough examples.

**Comment by esaulenka** [comment]
@vovaok, when you set offset=1 for returnaddress, did you see changes in Stack Editor dialog?
Looks like bug, that return address always located at zero offset.

**Comment by vovaok** [comment]
No, there is no difference. It looks like changing the offset of returnaddress is useless. If you see at the bottom of the Stack Editor there is **Return Address Offset** field. It always shows 0x0 regardless of settings in the .cspec file.
If you assign `byte` DataType to each stack frame entry it becomes clear what purpose has each byte of the frame with their default names (see picture below).
![image](https://user-images.githubusercontent.com/62897758/97041842-cf1d6100-1578-11eb-9839-2acf2853f7f4.png)
If I understand GHIDRA terminology correctly, the bytes at offsets 0x0 to 0x2 are considered as a result of the function, not even as return address. And I have no idea why there are 3 bytes, not 2. BTW this fact also doesn't depend on corresponding settings in .cspec.

**Comment by esaulenka** [comment]
> And I have no idea why there are 3 bytes, not 2

It is because arguments, passed through the stack, started at offset=3. It is stated in .cspec, and it works.
